Sources/RestBird-Alamofire/AlamofireSessionManager.swift

```diff
@@ -70,7 +70,31 @@ public final class AlamofireSessionManager: RestBird.SessionManager {
 extension AlamofireSessionManager {
 
     public func performUploadTask<Request: RestBird.UploadRequest>(request: Request, baseUrl: String, completion: @escaping (RestBird.Result<Data>) -> Void) {
-        let uploadRequest: Alamofire.UploadRequest
+        // We need to observe when `uploadRequest` gets set as in case of `.multipart` this will be set later, in `encodingCompletion` and we can't call these methods right after `sessionManager.upload` as `uploadRequest` will be nil at that point.
+        var uploadRequest: Alamofire.UploadRequest? {
+            didSet {
+                // We don't need to treat the case where `uploadRequest` is nil as that would be a failure and `completion(.failure(error))` would be called in `encodingCompletion`.
+                if let request = uploadRequest?.request {
+                    do {
+                        try delegate?.sessionManager(self, willPerform: request)
+                    } catch {
+                        completion(.failure(error))
+                    }
+                }
+                uploadRequest?.validate().responseData { response in
+                    if let urlRequest = response.request, let urlResponse = response.response {
+                        do {
+                            try self.delegate?.sessionManager(self, didPerform: urlRequest, response: urlResponse, data: response.data)
+                        } catch {
+                            completion(.failure(error))
+                            return
+                        }
+                    }
+                    completion(response.toResult())
+                }
+            }
+        }
+
         switch request.source {
```

Comment:
You have to handle the switch on the URLSessionManager as well

Reply:
I might be creating an issues for that for now as it is a bit harder to do this on URLSession and I don't think I have time for it this week.